### ISSUE TITLE
Actor Failure when killing Raylet

### ISSUE THREAD

**Issue by amogkam** [ray-project/xgboost_ray#14]
I am running simple.py with 700 workers, each with 1 CPU. I have 200 nodes/raylets in my cluster. I have `max_actor_restarts` set to 1. I start the script, and at the same time I run `ray kill-random-node` multiple times- which kills a random **raylet** (but doesn't terminate the node). After a few raylet kills, simple.py fails with this error:

```
2020-11-23 11:08:25,135	WARNING main.py:480 -- A Ray actor died during training. Trying to restart and continue training from last checkpoint.
Traceback (most recent call last):
  File "/root/simple.py", line 47, in <module>
    main()
  File "/root/simple.py", line 37, in main
    num_boost_round=100,)
  File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 475, in train
    **kwargs)
  File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 325, in _train
    checkpoint_frequency) for i in range(num_actors)
  File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 325, in <listcomp>
    checkpoint_frequency) for i in range(num_actors)
  File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 267, in _create_actor
    checkpoint_frequency=checkpoint_frequency), queue
  File "/root/anaconda3/lib/python3.7/site-packages/ray/actor.py", line 464, in remote
    override_environment_variables))
  File "/root/anaconda3/lib/python3.7/site-packages/ray/actor.py", line 673, in _remote
    or dict())
  File "python/ray/_raylet.pyx", line 1065, in ray._raylet.CoreWorker.create_actor
  File "python/ray/_raylet.pyx", line 1073, in ray._raylet.CoreWorker.create_actor
  File "python/ray/_raylet.pyx", line 144, in ray._raylet.check_status
ray.exceptions.GetTimeoutError: There was timeout in registering an actor. It is probably because GCS server is dead or there's a high load there.
```
Note that this all happens **before** actual training begins, so this is during actor creation/data loading. There is a warning message that gets printed out saying that an actor died during training, but then the script still errors out. I'm not sure whether subsequent kill-random-node calls killed the same actor that was restarting or a different one, but I was able to reproduce this error multiple times.

Also, encountering this issue, the Ray cluster becomes entirely unusable. I'm no longer able to connect to the Ray runtime and have to restart it to run anything else.

**Comment by richardliaw** [comment]
OK seems like a ray core issue. I think it's likely that:

1. This only will occur at start of training
2. This will only occur with multiple successive deaths

The next course of action is to transfer this issue to `ray-project/ray` and provide a repro script (or pseudocode -- the current text description of what you did to cause failure is a bit hard to follow).

**Comment by amogkam** [comment]
So I just tried again on master and I can't reproduce this. It is now correctly failing with 
```
Traceback (most recent call last):
  File "/root/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 856, in custom_excepthook
    ray.state.state.add_worker(worker_id, worker_type, worker_info)
  File "/root/anaconda3/lib/python3.7/site-packages/ray/state.py", line 733, in add_worker
    return self.global_state_accessor.add_worker_info(
AttributeError: 'NoneType' object has no attribute 'add_worker_info'

Original exception was:
Traceback (most recent call last):
  File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 579, in train
    **kwargs)
  File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 426, in _train
    ray.get(wait_load)
  File "/root/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 1377, in get
    raise value
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/simple.py", line 71, in <module>
    main(args.cpus_per_actor, args.num_actors)
  File "/root/simple.py", line 40, in main
    num_boost_round=100)
  File "/root/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 597, in train
    checkpoint_prefix))
RuntimeError: A Ray actor died during training and the maximum number of retries (1) is exhausted. Checkpoints have been stored at `/tmp/checkpoint/` with prefix `.xgb_ray_1606847841.142106` - you can pass these parameters as `checkpoint_path` and `checkpoint_prefix` to the `train()` function to try to continue the training.
```
